### PR TITLE
Fast style transfer example

### PR DESCRIPTION
**Pull request by vvmnnnkv**
Example model for issue #146

### CODE REVIEW

File: FastStyleTransfer/Layers/Helpers.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+
+/// Layer applying relu activation function
+public struct ReLU<Scalar: TensorFlowFloatingPoint>: Layer {
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return relu(input)
+    }
+}
```
Comment (rxwei) [inline FastStyleTransfer/Layers/Helpers.swift]: Leave a trailing line in every file.

File: FastStyleTransfer/Layers/Normalization.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import TensorFlow
+
+/// Layer that applies instance normalization over a mini-batch of images
+/// Expected input layout is BxHxWxC
+/// Reference: [Instance Normalization](https://arxiv.org/abs/1607.08022)
+public struct InstanceNorm2d<Scalar: TensorFlowFloatingPoint>: Layer {
```
Comment (rxwei) [inline FastStyleTransfer/Layers/Normalization.swift]:
```suggestion
public struct InstanceNorm2D<Scalar: TensorFlowFloatingPoint>: Layer {
```
`2d` -> `2D`. Same for other occurrences of `nD`.

File: FastStyleTransfer/Models/TransformerNet.swift [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import TensorFlow
+import Foundation
+
+/// Model that applies style
+public struct TransformerNet: Layer {
+    public typealias Input = Tensor<Float>
+    public typealias Output = Tensor<Float>
+
+    // conv layers
```
Comment (rxwei) [inline FastStyleTransfer/Models/TransformerNet.swift]:
```suggestion
    // Conv layers
```

File: FastStyleTransfer/Models/TransformerNet.swift [hunk 4]
```diff
@@ -0,0 +1,118 @@
+import TensorFlow
+import Foundation
+
+/// Model that applies style
```
Comment (rxwei) [inline FastStyleTransfer/Models/TransformerNet.swift]:
```suggestion
/// A model that applies style.
```
End every doc comment with a period.

File: FastStyleTransfer/Demo/Helpers.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        let parts = arg.split(separator: "=", maxSplits: 2)
+        let name = String(parts[0][parts[0].index(parts[0].startIndex, offsetBy: 2)...])
+        if let path = params[name], parts.count == 2 {
+            obj[keyPath: path] = String(parts[1])
+        }
+    }
+}
+
+enum FileError: Error {
+    case file_not_found
```
Comment (rxwei) [inline FastStyleTransfer/Demo/Helpers.swift]: Use camel case for all names.
```suggestion
    case fileNotFound
```

File: FastStyleTransfer/Demo/Helpers.swift [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import Foundation
+import TensorFlow
+import FastStyleTransfer
+
+// Make model importable
+extension TransformerNet: ImportableLayer {}
+
+func parseArgs<T>(into obj: inout T, with params: [String: WritableKeyPath<T, String?>]) {
```
Comment (rxwei) [inline FastStyleTransfer/Demo/Helpers.swift]: Avoid abbreviations.
```suggestion
func parseArguments<T>(into obj: inout T, with params: [String: WritableKeyPath<T, String?>]) {
```

File: FastStyleTransfer/Utility/Resize.swift [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import TensorFlow
+
+@differentiable(wrt: input, vjp: _vjpResizeNearestNeighbor(input:scale_factor:))
+public func resizeNearestNeighbor<Scalar: TensorFlowFloatingPoint>(_ input: Tensor<Scalar>, scale_factor: Float) -> Tensor<Scalar> {
+    let size = Tensor<Int32>(
+            shape: [2],
```
Comment (rxwei) [inline FastStyleTransfer/Utility/Resize.swift]: Use 4 space indentation.

File: FastStyleTransfer/Utility/Resize.swift [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import TensorFlow
+
+@differentiable(wrt: input, vjp: _vjpResizeNearestNeighbor(input:scale_factor:))
+public func resizeNearestNeighbor<Scalar: TensorFlowFloatingPoint>(_ input: Tensor<Scalar>, scale_factor: Float) -> Tensor<Scalar> {
```
Comment (rxwei) [inline FastStyleTransfer/Utility/Resize.swift]: Use camel case for all names including argument labels.

```suggestion
public func resizeNearestNeighbor<Scalar: TensorFlowFloatingPoint>(_ input: Tensor<Scalar>, scaleFactor: Float) -> Tensor<Scalar> {
```

File: FastStyleTransfer/Utility/ImportableLayer.swift [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+/// Adds ability to recursively override Layer properties
+/// WARNING This is temporary solution it relies on the fact that Mirror and KeyPathIterable enumerate properties in the same order
+
+public protocol ImportableLayer: KeyPathIterable {}
+
+public typealias ImportMap = [String: (String, [Int]?)]
+
+public extension ImportableLayer {
+
```
Comment (rxwei) [inline FastStyleTransfer/Utility/ImportableLayer.swift]: Remove redundant empty line.

File: FastStyleTransfer/Layers/Helpers.swift (hunk 1 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Layers/Helpers.swift]: Done

File: FastStyleTransfer/Layers/Normalization.swift (hunk 2 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Layers/Normalization.swift]: Done

File: FastStyleTransfer/Models/TransformerNet.swift (hunk 3 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Models/TransformerNet.swift]: Done

File: FastStyleTransfer/Models/TransformerNet.swift (hunk 4 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Models/TransformerNet.swift]: Done

File: FastStyleTransfer/Demo/Helpers.swift (hunk 5 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Demo/Helpers.swift]: Done

File: FastStyleTransfer/Demo/Helpers.swift (hunk 6 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Demo/Helpers.swift]: Done

File: FastStyleTransfer/Utility/Resize.swift (hunk 7 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Utility/Resize.swift]: Done

File: FastStyleTransfer/Utility/Resize.swift (hunk 8 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Utility/Resize.swift]: Done

File: FastStyleTransfer/Utility/ImportableLayer.swift (hunk 9 above)
Comment (vvmnnnkv) [reply FastStyleTransfer/Utility/ImportableLayer.swift]: Done

**Comment by vvmnnnkv** [comment]
Hi @rxwei, I've tried to spot & fix similar issues manually, but do you use some kind of 'official' linter to make this job automatic?

**Comment by BradLarson** [comment]
@vvmnnnkv - For formatting, I use `swift-format`: https://github.com/apple/swift-format , for which we have [a .swift-format file](https://github.com/tensorflow/swift-models/blob/master/.swift-format) in this repository that defines the formatting conventions used here. It really helps to catch even subtle formatting issues.

Might not work as well on notebooks, however.

**Comment by vvmnnnkv** [comment]
@BradLarson - interesting, https://github.com/apple/swift-format crashes when compiled in Ubuntu (with DEVELOPMENT-SNAPSHOT-2019-07-10-a). I'm not sure if this repo is the right `swift-format` @rxwei mentions in this PR https://github.com/tensorflow/swift-apis/pull/374? I also found  https://github.com/google/swift/tree/format but it looks outdated.
Compiling `apple/swift-format` with Xcode 11b4 works, though. I've applied it to the code, but some places seem to look less readable after that, e.g. inline closure here: https://github.com/tensorflow/swift-models/pull/191/commits/29e9f88ea4303fdd872659a08c767562689c5ebc#diff-0a85214f52c515dacd5ce305c43b1720

**Comment by BradLarson** [comment]
@vvmnnnkv - Try building the public swift-format using one of the standard Swift toolchains, not a Swift for TensorFlow toolchain. I saw similar problems on my end using an S4TF toolchain a while ago. I use a swift-format that I built using a public 5.1 toolchain.

**Comment by BradLarson** [comment]
@vvmnnnkv - Yes, I've seen similar problems with Swift for TensorFlow toolchains vs. stock Swift toolchains on swift-format. I think it's something that's being looked into. In the meantime, using swift-format with the stock toolchain should be fine.

We do know that it has some problems around a few specific cases, so you may need to hand-correct in a couple of places. It works well overall, but isn't perfect. I know those cases are being worked on.

**Comment by vvmnnnkv** [comment]
@rxwei @BradLarson 
I think I'm done with all changes.

File: FastStyleTransfer/Layers/Normalization.swift [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer. Expected input layout is BxHxWxC.
+    /// - Returns: The output.
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        // Calculate mean & variance along H,W axes.
+        let mean = input.mean(alongAxes: [1, 2])
```
Comment (t-ae) [inline FastStyleTransfer/Layers/Normalization.swift]: `moments` computes `mean` and `variance` simultaneously.
https://github.com/tensorflow/swift-apis/blob/c7595c4b4b0e824cd6d9449e2bed39944fafb472/Sources/TensorFlow/Operators/Math.swift#L2501-L2512

File: FastStyleTransfer/Layers/Normalization.swift [hunk 11]
```diff
@@ -0,0 +1,37 @@
+import TensorFlow
+
+/// 2-D layer applying instance normalization over a mini-batch of inputs.
+///
+/// Reference: [Instance Normalization](https://arxiv.org/abs/1607.08022)
+public struct InstanceNorm2D<Scalar: TensorFlowFloatingPoint>: Layer {
```
Comment (t-ae) [inline FastStyleTransfer/Layers/Normalization.swift]: How about adding `InstanceNorm2D` to swift-apis?

**Comment by BradLarson** [comment]
Sorry it has been so long since we've reviewed this. I quick resolved the couple of merge conflicts, but it looks like a few areas of this model might need to be brought up to date to get this to build under the current toolchain.

The warnings don't look too hard to fix (Raw has been renamed to _Raw, `transposed(permutation:)` instead of `transposed(withPermutations:)`). I think you might need to make ReflectionPad2D and maybe others ParameterlessLayers to resolve the `...does not conform to protocol 'VectorProtocol'` errors, but I might be mistaken there.

If you can get this to build cleanly under the current toolchain, I think this would be good to pull in. It looks like all previous comments have been addressed.

**Comment by vvmnnnkv** [comment]
Hi @BradLarson 
With suggested changes it compiles fine on v0.6 toolchain (Ubuntu). I also removed tensor extension for reflected padding because it's now included in `swift-apis`. Perhaps, I did this too early because example notebook doesn't work with Colab after that (error: there's no `mode` param in `Tensor.padded`). Do you know if Colab is going to be updated to v0.6 soon?

**Comment by BradLarson** [comment]
We're looking into the test failure, which looks to be a toolchain artifact. Once we get that resolved, I'll re-run CI and make sure everything's green. Sorry for the delay due to the U.S. holiday.

**Comment by BradLarson** [comment]
We think we have a workaround for the failing test. Could you bring this branch up to date with the head of swift-models? I think that should allow for the tests to pass, with Marc's workaround that he committed yesterday.

**Pull request merged**